improve benchmarks

alexlamsl opened the pull request:
The major change is with `benchmark.js`:
- `gzip` is not always available (e.g. Windows), so use Node.js `zlib` instead
- incorporate `Minimize` results into the table, also apply `gzip` to its output

Minor changes are updating `devDependencies` and restoring a few commented-out tests.

alexlamsl committed:
update dependencies

alexlamsl committed:
restore disabled tests

alexlamsl committed:
improve benchmark.js
- use internal zlib instead of external gzip command
- perform gzip on Minimized output
- incorporate Minimize results into table

alexlamsl commented:
One thing that I'm interested in doing it is to run [Will Peavy](http://www.willpeavy.com/minifier/) and [htmlcompressor.com](http://htmlcompressor.com/) off `benchmark.js` as well and end up with a big table like the one on the front page.

But I'm not sure if it's orthodox to ping those sites in a script like that, so I would like a second opinion on the matter :sweat_smile:

alexlamsl commented:
| File | Original | HTMLMinifier | Minimize | WillPeavy | htmlcompressor | Savings |
| --- | --: | --: | --: | --: | --: | --: |
| abc | 91.00 | 77.98 | 87.05 | 85.83 | 84.83 | 14.31% |
| amazon | 245.93 | 211.50 | 240.29 | 230.49 | 219.20 | 14.00% |
| eloquentjavascript | 869.53 | 830.58 | 838.30 | 864.16 |  | 4.48% |
| es6 | 3677.55 | 3060.28 | 3169.03 | 3204.39 |  | 16.78% |
| es6-draft | 3619.61 | 3006.00 | 3115.87 | 3140.62 |  | 16.95% |
| es6-table | 117.94 | 80.12 | 109.65 | 94.70 | 94.22 | 32.07% |
| google | 133.13 | 128.17 | 132.50 | 135.37 | 131.38 | 3.72% |
| html-minifier | 48.81 | 38.24 | 43.86 | 43.97 | 42.63 | 21.65% |
| msn | 156.69 | 133.97 | 137.94 | 144.92 | 138.37 | 14.50% |
| newyorktimes | 131.41 | 110.74 | 123.15 | 124.55 | 119.64 | 15.73% |
| stackoverflow | 200.47 | 160.35 | 184.92 | 173.87 | 166.29 | 20.01% |
| wikipedia | 401.42 | 381.44 | 390.04 | 400.40 |  | 4.98% |

alexlamsl commented:
gzipped:

| File | Original | HTMLMinifier | Minimize | WillPeavy | htmlcompressor | Savings |
| --- | --: | --: | --: | --: | --: | --: |
| abc | 16.69 | 14.67 | 16.81 | 16.11 | 16.00 | 12.06% |
| amazon | 56.17 | 49.75 | 55.33 | 53.27 | 50.91 | 11.42% |
| eloquentjavascript | 196.96 | 188.25 | 191.18 | 192.92 |  | 4.42% |
| es6 | 424.58 | 373.83 | 390.23 | 379.48 |  | 11.95% |
| es6-draft | 425.84 | 374.39 | 391.70 | 380.88 |  | 12.08% |
| es6-table | 5.89 | 5.21 | 6.75 | 5.54 | 5.49 | 11.68% |
| google | 39.25 | 38.17 | 39.15 | 39.17 | 38.85 | 2.74% |
| html-minifier | 9.95 | 8.77 | 9.66 | 9.32 | 9.25 | 11.93% |
| msn | 41.22 | 37.18 | 37.73 | 37.95 | 38.03 | 9.80% |
| newyorktimes | 29.25 | 25.59 | 28.14 | 27.48 | 27.08 | 12.51% |
| stackoverflow | 31.86 | 30.05 | 31.86 | 30.76 | 30.59 | 5.68% |
| wikipedia | 80.11 | 78.84 | 80.55 | 79.80 |  | 1.58% |

kangax commented:
Awesome work @alexlamsl! 

Couple things:
1. It's probably best to combine gzip results right into 1 table (with ungzipped results)
2. es6 and es6-draft are the same thing, IIRC so let's ditch the draft one
3. I've been meaning to add Zopfli and Brotli so this might be a good time for it
4. Re: pinging site — I'm all for unorthodox solutions so totally go for it :)

alexlamsl commented:
I haven't touched the output table format with regards to `+ gzipped` - I merely separated them by hand when I post the results here :sweat_smile:

kangax commented:
Oh good. Sorry, can't keep track of all my open source projects :D

alexlamsl committed:
rearrange options in alphabetical order

alexlamsl committed:
added Will Peavy and htmlcompressor.com

alexlamsl committed:
removed es6-draft

alexlamsl commented:
Added the two aforementioned tools for comparison, removed `es6-draft`.

By "Zopfli and Brotli" do you mean using them as opposed to `zlib`'s standard `gzip` routines to see if there's any difference in space savings?

alexlamsl committed:
insert some comments

kangax commented:
> By "Zopfli and Brotli" do you mean using them as opposed to zlib's standard gzip routines to see if there's any difference in space savings?

Yes, to see how different compression algorithms work on our minified code. From what I understand, repeating chunks in text are "good" for compression algorithms like gzip. And I also recall seeing some Zopfli benchmarks where _larger_ files had better compression results than smaller ones. Given all that, it would be interesting to see overall picture of which savings happen when, with which algorithms, and to which extent.

kangax commented:
For example, imagine if Brotli (the future compression standard) works so good that our minification becomes unnecessary. Then our work here is done :)

kangax commented on benchmark.js:
```diff
@@ -55,89 +56,284 @@ function blueTime(time) {
 
-          // Open and read the size of the minified output
-          fs.stat(minifiedFilePath, function (err, stats) {
+    run([
```
Ok, that's one big &$% call expression :) Best to refactor.

kangax commented on benchmark.js:
```diff
@@ -55,89 +56,284 @@ function blueTime(time) {
+              }, function () {
+                run([
+                  // Gzip the minified output
+                  function (done) {
```
This is all duplicated across tests, right? Definitely need to DRY it up.

kangax commented on tests/minifier.js:
```diff
@@ -1008,22 +1008,21 @@
-    //   removeCommentsFromCDATA: true,
-    //   minifyJS: true
-    // }), output);
+    input = '<script>\n' +
```
:+1: thanks :)

alexlamsl commented:
Rumour has it Skynet is the ultimate answer :laughing: 

Jokes aside, I am looking for pure JavaScript implementations for both as it's most portable IMHO. And so far Emscripten versions seem to be the safest bet.
- [Brotli](https://www.npmjs.com/package/brotli)
- [Zopfli](https://github.com/imaya/zopfli.js)

alexlamsl committed:
refactor functions

alexlamsl committed:
improve code reuse

alexlamsl commented:
@kangax addressed your comments, PTAL.

alexlamsl commented:
Random thought: shall we go fancy with [progress](https://www.npmjs.com/package/progress)?

Those `processing <page>...` don't look too informative atm...

alexlamsl committed:
add progress bar

kangax commented on benchmark.js:
```diff
@@ -55,89 +61,227 @@ function blueTime(time) {
-                  toKb(data.length),
-                  toKb(minifiedSize));
+    function testWillPeavy(done) {
+      fs.readFile(filePath, {
```
Definitely looks much better, thanks! One final nit though is that we can probably still cut down on repetition a bit more here. There's all this similar boilerplate with reading file, throwing error, setting headers and so on, which we can extract into a generic method. It will help with accidentally changing something in one location (e.g. testWillPeavy) and forgetting to change another. And it will make it much easier to test other pages if the need arises. Or is there too much differences between them?

alexlamsl commented:
| function call | frequency |
| --- | --- |
| fs.writeFile(Buffer) | 1 |
| fs.writeFile(String) | 2 |
| fs.readFile(Buffer) | 1 |
| fs.readFile(String) | 2 |

So I'm not sure if extracting any of those would help with readability. In addition, I can forsee adjusting these site tests independently from one another, i.e. one site changes their layout.

alexlamsl commented:
So I'm testing `brotli` - it's taking quite a while to `compress()`, and I think it'd be better to `child_process.fork()` it rather than blocking the main thread.

Question: given `benchmark.js` is in root directory, where should I put this `benchmark-brotli.js`?

kangax commented:
Root seems fine. Otherwise gotta move both under benchmarks.

Sent from my iPhone

> On 02 Feb 2016, at 00:05, Alex Lam S.L. notifications@github.com wrote:
> 
> So I'm testing brotli - it's taking quite a while to compress(), and I think it'd be better to child_process.fork() it rather than blocking the main thread.
> 
> Question: given benchmark.js is in root directory, where should I put this benchmark-brotli.js?
> 
> —
> Reply to this email directly or view it on GitHub.

kangax commented:
Fair enough, let's leave these as is.

Sent from my iPhone

> On 01 Feb 2016, at 22:59, Alex Lam S.L. notifications@github.com wrote:
> 
> function call frequency
> fs.writeFile(Buffer)  1
> fs.writeFile(String)  2
> fs.readFile(Buffer)   1
> fs.readFile(String)   2
> So I'm not sure if extracting any of those would help with readability. In addition, I can forsee adjusting these site tests independently from one another, i.e. one site changes their layout.
> 
> —
> Reply to this email directly or view it on GitHub.

alexlamsl committed:
add brotli to benchmark

alexlamsl commented:
Right, added brotli compression to all minification tests. Done everything sequentially to avoid socket time-outs and make timing measurements more accurate.

alexlamsl commented:
`zopfli.js` isn't on npm, and the zopfli-related packages I can find on npm all requires installing the native binaries :cry:

alexlamsl commented:
![benchmark](https://cloud.githubusercontent.com/assets/1898413/12741888/74edc23c-c9bb-11e5-9485-0591c54ef05f.png)

alexlamsl committed:
add LZMA to benchmark

alexlamsl commented:
![benchmark](https://cloud.githubusercontent.com/assets/1898413/12749130/9c299c1c-c9e9-11e5-92ee-958718a98fdb.png)
Fun fact: these are all run on [node-chakracore](https://github.com/Microsoft/node/releases)

alexlamsl commented:
Looking around for other tools to compare with, and found [some](https://www.npmjs.com/package/node-minifier) peculiar [ones](https://www.npmjs.com/package/html-minify).

alexlamsl commented:
Okay, found [an interesting one](http://www.minifyweb.com/minify-html/)

kangax commented:
Hm, is something wrong with our savings calculations?

Take a look:

![screenshot 2016-02-02 16 54 38](https://cloud.githubusercontent.com/assets/383/12765655/acbace8c-c9cd-11e5-8325-86667fd3b93a.png)

kangax commented:
Looks like "html-minify" is just a bundled version of ours — https://github.com/yize/html-minify/blob/master/lib/minify.js#L350-L357

alexlamsl commented:
The savings look legit to me - the total size is smaller, so even with smaller absolute savings the percentage is not necessarily going to be smaller...
